OpenRA.Mods.Common/Traits/GainsExperience.cs

```diff
@@ -77,32 +77,35 @@ public GainsExperience(ActorInitializer init, GainsExperienceInfo info)
 				nextLevel.Add(Pair.New(kv.Key * cost, kv.Value));
 
 			if (init.Contains<ExperienceInit>())
-				GiveExperience(init.Get<ExperienceInit, int>());
+				GiveExperience(init.Get<ExperienceInit, int>(), true);
 		}
 
 		public bool CanGainLevel { get { return Level < MaxLevel; } }
 
-		public void GiveLevels(int numLevels)
+		public void GiveLevels(int numLevels, bool silent = false)
 		{
 			var newLevel = Math.Min(Level + numLevels, MaxLevel);
-			GiveExperience(nextLevel[newLevel - 1].First - experience);
+			GiveExperience(nextLevel[newLevel - 1].First - experience, silent);
 		}
 
-		public void GiveExperience(int amount)
+		public void GiveExperience(int amount, bool silent = false)
 		{
 			experience += amount;
 
+			var um = self.Trait<UpgradeManager>();
```

Comment:
As far as I know, `GainsExperience` doesn't work without upgrades (`UpgradeManager`) at all, so that would make it silently fail.
Declare the dependency explicitly with `Requires<T>`.